From f4523ddf6e32b852065dfc5dbe25ba875b0df849 Mon Sep 17 00:00:00 2001
From: Alan Tull <r80115@freescale.com>
Date: Mon, 8 Sep 2008 13:28:31 -0500
Subject: [PATCH] ENGR00070691-2 mx51: add audio support

Add audio playback for i.Mx51 using Wolfson wm8903 codec
on CPU3 (AAPL) board.
Port wm8903 codec driver from Wolfson GIT repository.
http://opensource.wolfsonmicro.com/node/8
IRAM support not included.
Mono support not included.
16 and 24 bit playback tested to be working.

Signed-off-by: Alan Tull <r80115@freescale.com>
---
 arch/arm/configs/imx51_3stack_defconfig |   60 +++++++++++++++++++++++++++++-
 arch/arm/mach-mx51/devices.c            |   27 ++++++++++++++
 arch/arm/mach-mx51/mx51_3stack.c        |    9 +++--
 arch/arm/mach-mx51/mx51_3stack_gpio.c   |   36 +++++++++++++++++--
 include/asm-arm/arch-mxc/mx51.h         |   15 ++++++++
 5 files changed, 139 insertions(+), 8 deletions(-)

diff --git a/arch/arm/configs/imx51_3stack_defconfig b/arch/arm/configs/imx51_3stack_defconfig
index fdf069c..d46b799 100644
--- a/arch/arm/configs/imx51_3stack_defconfig
+++ b/arch/arm/configs/imx51_3stack_defconfig
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.24
-# Tue Aug 19 12:07:39 2008
+# Wed Aug 20 11:18:20 2008
 #
 CONFIG_ARM=y
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
@@ -889,7 +889,63 @@ CONFIG_LOGO_LINUX_CLUT224=y
 #
 # Sound
 #
-# CONFIG_SOUND is not set
+CONFIG_SOUND=y
+
+#
+# Advanced Linux Sound Architecture
+#
+CONFIG_SND=y
+CONFIG_SND_TIMER=y
+CONFIG_SND_PCM=y
+# CONFIG_SND_SEQUENCER is not set
+CONFIG_SND_OSSEMUL=y
+CONFIG_SND_MIXER_OSS=y
+CONFIG_SND_PCM_OSS=y
+CONFIG_SND_PCM_OSS_PLUGINS=y
+# CONFIG_SND_DYNAMIC_MINORS is not set
+CONFIG_SND_SUPPORT_OLD_API=y
+CONFIG_SND_VERBOSE_PROCFS=y
+# CONFIG_SND_VERBOSE_PRINTK is not set
+# CONFIG_SND_DEBUG is not set
+
+#
+# Generic devices
+#
+# CONFIG_SND_DUMMY is not set
+# CONFIG_SND_MTPAV is not set
+# CONFIG_SND_SERIAL_U16550 is not set
+# CONFIG_SND_MPU401 is not set
+
+#
+# ALSA ARM devices
+#
+# CONFIG_SND_MXC_SPDIF is not set
+
+#
+# SPI devices
+#
+
+#
+# System on Chip audio support
+#
+CONFIG_SND_SOC=y
+
+#
+# SoC Audio support for SuperH
+#
+CONFIG_SND_MXC_SOC=y
+CONFIG_SND_MXC_SOC_SSI=y
+# CONFIG_SND_MXC_SOC_IRAM is not set
+# CONFIG_SND_SOC_IMX_3STACK_WM8350 is not set
+# CONFIG_SND_SOC_IMX_3STACK_AK4647 is not set
+# CONFIG_SND_SOC_IMX_3STACK_WM8580 is not set
+CONFIG_SND_SOC_IMX_3STACK_WM8903=y
+CONFIG_SND_SOC_WM8903=y
+
+#
+# Open Sound System
+#
+# CONFIG_SOUND_PRIME is not set
 # CONFIG_HID_SUPPORT is not set
 CONFIG_USB_SUPPORT=y
 
diff --git a/arch/arm/mach-mx51/devices.c b/arch/arm/mach-mx51/devices.c
index 5bb2e4f..5a63260 100644
--- a/arch/arm/mach-mx51/devices.c
+++ b/arch/arm/mach-mx51/devices.c
@@ -557,6 +557,32 @@ static inline void mxc_init_dma(void)
 	(void)platform_device_register(&mxc_dma_device);
 }
 
+static struct mxc_audio_platform_data mxc_audio_data;
+
+static struct platform_device mxc_alsa_device = {
+	.name = "imx-3stack-wm8903",
+	.id = 0,
+	.dev = {
+		.release = mxc_nop_release,
+		.platform_data = &mxc_audio_data,
+		},
+};
+
+static void mxc_init_audio(void)
+{
+	mxc_audio_data.ssi_clk[0] = clk_get(NULL, "ssi_clk.0");
+	clk_put(mxc_audio_data.ssi_clk[0]);
+
+	mxc_audio_data.ssi_clk[1] = clk_get(NULL, "ssi_clk.1");
+	clk_put(mxc_audio_data.ssi_clk[1]);
+
+	mxc_audio_data.ssi_num = 1;
+	mxc_audio_data.src_port = 2;
+	mxc_audio_data.ext_port = 3;
+
+	platform_device_register(&mxc_alsa_device);
+}
+
 static int __init mxc_init_devices(void)
 {
 	mxc_init_wdt();
@@ -567,6 +593,7 @@ static int __init mxc_init_devices(void)
 	mxc_init_owire();
 	mxc_init_ipu();
 	mxc_init_vpu();
+	mxc_init_audio();
 
 	return 0;
 }
diff --git a/arch/arm/mach-mx51/mx51_3stack.c b/arch/arm/mach-mx51/mx51_3stack.c
index c1fc7e1..697e255 100644
--- a/arch/arm/mach-mx51/mx51_3stack.c
+++ b/arch/arm/mach-mx51/mx51_3stack.c
@@ -265,7 +265,7 @@ static inline void mxc_init_bl(void)
 	platform_device_register(&mxcbl_device);
 }
 
-#ifdef CONFIG_I2C_MXC
+#if defined(CONFIG_I2C_MXC) || defined(CONFIG_I2C_MXC_MODULE)
 
 #ifdef CONFIG_I2C_MXC_SELECT1
 static struct i2c_board_info mxc_i2c0_board_info[] __initdata = {
@@ -279,7 +279,10 @@ static struct i2c_board_info mxc_i2c1_board_info[] __initdata = {
 	 .addr = 0x08,
 	 .platform_data = (void *)MX51_PIN_GPIO1_5,
 	 },
-
+	{
+	 .type = "wm8903-i2c",
+	 .addr = 0x1a,
+	 },
 };
 #endif
 
@@ -608,7 +611,7 @@ static void __init mxc_board_init(void)
 	mxc_init_nand_mtd();
 	mxc_init_mmc();
 
-#ifdef CONFIG_I2C_MXC
+#if defined(CONFIG_I2C_MXC) || defined(CONFIG_I2C_MXC_MODULE)
 
 #ifdef CONFIG_I2C_MXC_SELECT1
 	i2c_register_board_info(0, mxc_i2c0_board_info,
diff --git a/arch/arm/mach-mx51/mx51_3stack_gpio.c b/arch/arm/mach-mx51/mx51_3stack_gpio.c
index 2e2c596..31fc0d3 100644
--- a/arch/arm/mach-mx51/mx51_3stack_gpio.c
+++ b/arch/arm/mach-mx51/mx51_3stack_gpio.c
@@ -376,6 +376,7 @@ void gpio_i2c_active(int i2c_num)
 				  PAD_CTL_HYS_ENABLE |
 				  PAD_CTL_DRV_VOT_LOW | PAD_CTL_DDR_INPUT_CMOS);
 		break;
+
 	case 1:
 		mxc_request_iomux(MX51_PIN_GPIO1_2,
 				  IOMUX_CONFIG_ALT2 | IOMUX_CONFIG_SION);
@@ -401,7 +402,6 @@ void gpio_i2c_active(int i2c_num)
 				  PAD_CTL_100K_PU |
 				  PAD_CTL_HYS_ENABLE |
 				  PAD_CTL_DRV_VOT_LOW | PAD_CTL_DDR_INPUT_CMOS);
-
 		break;
 	case 2:
 		break;
@@ -473,11 +473,41 @@ void gpio_pmic_active(void)
 EXPORT_SYMBOL(gpio_pmic_active);
 
 /*!
- * This function activates DAM ports 4 & 5 to enable
- * audio I/O.
+ * This function activates DAM port 3 to enable audio I/O.
  */
 void gpio_activate_audio_ports(void)
 {
+	unsigned int pad_val;
+
+	pad_val = PAD_CTL_SRE_FAST | PAD_CTL_DRV_HIGH |
+	    PAD_CTL_ODE_OPENDRAIN_NONE | PAD_CTL_100K_PU |
+	    PAD_CTL_HYS_NONE | PAD_CTL_DDR_INPUT_CMOS | PAD_CTL_DRV_VOT_LOW;
+
+	/* AUD3_TXD */
+	mxc_request_iomux(MX51_PIN_AUD3_BB_TXD, IOMUX_CONFIG_ALT0);
+	mxc_iomux_set_pad(MX51_PIN_AUD3_BB_TXD, pad_val);
+
+	/* AUD3_RXD */
+	mxc_request_iomux(MX51_PIN_AUD3_BB_RXD, IOMUX_CONFIG_ALT0);
+	mxc_iomux_set_pad(MX51_PIN_AUD3_BB_RXD, pad_val);
+
+	/* AUD3_CLK */
+	mxc_request_iomux(MX51_PIN_AUD3_BB_CK, IOMUX_CONFIG_ALT0);
+	mxc_iomux_set_pad(MX51_PIN_AUD3_BB_CK, pad_val);
+
+	/* AUD3_FS */
+	mxc_request_iomux(MX51_PIN_AUD3_BB_FS, IOMUX_CONFIG_ALT0);
+	mxc_iomux_set_pad(MX51_PIN_AUD3_BB_FS, pad_val);
+
+	/*
+	 * CPU3 AAPL board:
+	 * codec mclk is driven by 12MHz external xtal oscillator.
+	 * enable it by driving EIM_D16 high.
+	 */
+	mxc_request_iomux(MX51_PIN_EIM_D16, IOMUX_CONFIG_ALT1);
+	mxc_iomux_set_pad(MX51_PIN_EIM_D16, pad_val);
+	mxc_set_gpio_direction(MX51_PIN_EIM_D16, 0);
+	mxc_set_gpio_dataout(MX51_PIN_EIM_D16, 1);
 }
 
 EXPORT_SYMBOL(gpio_activate_audio_ports);
diff --git a/include/asm-arm/arch-mxc/mx51.h b/include/asm-arm/arch-mxc/mx51.h
index b15cdd4..6b07a28 100644
--- a/include/asm-arm/arch-mxc/mx51.h
+++ b/include/asm-arm/arch-mxc/mx51.h
@@ -69,6 +69,21 @@
  */
 #define IRAM_BASE_ADDR		0x1FFE8000	/* internal ram */
 
+#ifdef CONFIG_SDMA_IRAM
+#define SDMA_IRAM_SIZE  CONFIG_SDMA_IRAM_SIZE
+#else
+#define SDMA_IRAM_SIZE  0
+#endif
+
+#ifdef CONFIG_SND_MXC_SOC_IRAM
+#define SND_RAM_SIZE 0x6000
+#else
+#define SND_RAM_SIZE 0
+#endif
+
+#define SDMA_RAM_BASE_ADDR (IRAM_BASE_ADDR)
+#define SND_RAM_BASE_ADDR	(IRAM_BASE_ADDR + SDMA_IRAM_SIZE)
+
 /*
  * NFC
  */
-- 
1.5.4.4

